library/src/main/java/SpecTreesUnionizer.java

```diff
@@ -0,0 +1,256 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+
+/** Provides functions for performing union operations on spec trees represented as Maps. */
+public class SpecTreesUnionizer {
+  /**
+   * Performs a union on {@code map1} and {@code map2} and returns the result.
+   *
+   * @param map1 the map which {@code map2} is merged into. {@code map1} will contain the result of
+   *     the union.
+   * @param map2 the map to merge into {@code map1}
+   * @return the result of a union on {@code map1} and {@code map2}
+   */
+  static LinkedHashMap<String, Object> union(
+      LinkedHashMap<String, Object> map1, LinkedHashMap<String, Object> map2)
+      throws UnionConflictException, UnexpectedTypeException {
+    UnionizerUnionParams unionizerUnionParams = UnionizerUnionParams.builder().build();
+
+    return union(map1, map2, unionizerUnionParams);
+  }
+
+  /**
+   * Performs a union on {@code map1} and {@code map2} with the given {@code unionizerUnionParams}
+   * and returns the result.
+   *
+   * @param map1 the map which {@code map2} is merged into. {@code map1} will contain the result of
+   *     the union.
+   * @param map2 the map to merge into {@code map1}
+   * @param unionizerUnionParams a set of special options which can be applied during the union
+   * @return the result of a union on {@code map1} and {@code map2} with options from {@code
+   *     unionizerUnionParams} applied
+   */
+  static LinkedHashMap<String, Object> union(
+      LinkedHashMap<String, Object> map1,
```

Comment:
I got a little lost reading thru this and I think I came back to the variable names were a bit confusing.   I am not sure how to fix it but something more telling of as mapToMergeInto or mapToMerge.   I understood what you are doing but it took some time to wrap my head around it and I am thinking from a readibility standpoint from an opensource user.   I am not saying my names are better, but just trying to think of something which may help people understand quicker.   Maybe some examples?

Reply:
+1

Reply:
always opt for clear names (instead of map1, map2, etc)

Reply:
addressed!